game-core/src/main/java/games/strategy/triplea/odds/calculator/ConcurrentBattleCalculator.java

```diff
@@ -223,44 +223,41 @@ public AggregateResults calculate(
       final Collection<Unit> bombarding,
       final Collection<TerritoryEffect> territoryEffects,
       final boolean retreatWhenOnlyAirLeft,
-      final int initialRunCount)
+      final int runCount)
       throws IllegalStateException {
     synchronized (mutexCalcIsRunning) {
       awaitLatch();
       final long start = System.currentTimeMillis();
+      final AtomicInteger overflowRunsAvailable = new AtomicInteger(runCount % workers.size());
       // Create worker thread pool and start all workers
-      int totalRunCount = 0;
-      int runCount = initialRunCount;
-      final int workerNum = workers.size();
-      final int workerRunCount = Math.max(1, (runCount / Math.max(1, workerNum)));
-      final List<Future<AggregateResults>> list = new ArrayList<>();
-      for (final BattleCalculator worker : workers) {
-        if (!getIsReady()) {
-          // we could have attempted to set a new game data, while the old one was still being set,
-          // causing it to abort with null data
-          return new AggregateResults(0);
-        }
-        final int currentWorkedRunCount = (runCount <= 0 ? 0 : workerRunCount);
-        if (currentWorkedRunCount > 0) {
-          totalRunCount += currentWorkedRunCount;
-          list.add(
-              executor.submit(
-                  () ->
-                      worker.calculate(
-                          attacker,
-                          defender,
-                          location,
-                          attacking,
-                          defending,
-                          bombarding,
-                          territoryEffects,
-                          retreatWhenOnlyAirLeft,
-                          currentWorkedRunCount)));
-        }
-        runCount -= workerRunCount;
+      final int runsPerWorker = runCount / workers.size();
+      final List<Future<AggregateResults>> list =
+          workers.stream()
+              .map(
+                  worker ->
+                      executor.submit(
+                          () ->
+                              worker.calculate(
+                                  attacker,
+                                  defender,
+                                  location,
+                                  attacking,
+                                  defending,
+                                  bombarding,
+                                  territoryEffects,
+                                  retreatWhenOnlyAirLeft,
+                                  // Ensure that we always achieve the target run count even if
+                                  // the number is not dividable by workers.size()
+                                  (overflowRunsAvailable.getAndDecrement() > 0 ? 1 : 0)
+                                      + runsPerWorker)))
+              .collect(Collectors.toList());
```

Comment:
suggestion: What are your thoughts to add a postcondition here to check that sum of aggregated results is equal to the run count?

Reply:
Not quite sure if that would make sense.
I mean in the end it doesn't really matter if we get exactly the number of runs we required or just something that's very similar. The only "real" reason that would cause a significant difference in runs would be if a worker got an exception, which should be caught by our error handling later anyways.

So in the end if I make the "distribution logic" its own stateful object that distributes run counts evenly we could just add unit tests for that and that should make any potential post-condition redundant